src/transaction/buffer.rs

```diff
@@ -8,13 +8,42 @@ use std::{
 use tikv_client_proto::kvrpcpb;
 use tokio::sync::Mutex;
 
+#[derive(Default)]
+struct Mutations {
+    primary_key: Option<Key>,
+    key_mutation_map: BTreeMap<Key, Mutation>,
+}
+
+impl Mutations {
+    fn insert(&mut self, key: impl Into<Key>, mutation: Mutation) {
+        let key = key.into();
+        if self.primary_key.is_none() {
+            self.primary_key = Some(key.clone());
+        }
+        self.key_mutation_map.insert(key, mutation);
+    }
+
+    pub fn get_primary_key_or(&mut self, key: Key) -> Key {
+        self.primary_key.get_or_insert(key).clone()
+    }
+}
+
 /// A caching layer which buffers reads and writes in a transaction.
 #[derive(Default)]
 pub struct Buffer {
-    mutations: Mutex<BTreeMap<Key, Mutation>>,
+    mutations: Mutex<Mutations>,
 }
 
 impl Buffer {
+    /// Get the primary key of the buffer, if not exists, use `key` as the primary key.
+    pub async fn get_primary_key(&self) -> Option<Key> {
+        self.mutations.lock().await.primary_key.clone()
+    }
+    /// Get the primary key of the buffer, if not exists, use `key` as the primary key.
+    pub async fn get_primary_key_or(&self, key: Key) -> Key {
```

Comment:
Can we take the key by reference rather than value so the caller doesn't need to clone, then only clone if there is no primary key?

Reply:
Also return a reference so the caller can clone

Reply:
We do can pass the param by reference, but it's hard to return a reference here due to the mutex.